fs/vfs/fs_chmod.c

```diff
@@ -0,0 +1,103 @@
+/****************************************************************************
+ * fs/vfs/fs_chmod.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+
+#include <errno.h>
+#include <assert.h>
+
+#include <nuttx/fs/fs.h>
+
+#include "inode/inode.h"
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: chown
+ *
+ * Description:
+ *   changes the mode of the file specified whose pathname is given in
+ *   pathname, which is dereferenced if it is a symbolic link.
+ *
+ * NOTE: Due to the characteristics of fs, this function is incomplete.
+ *
+ * Input Parameters:
+ *   fd    : fd.
+ *   mode  : Permission value.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success; -1 is returned on fail, and errno is
+ *   set appropriately.
+ *
+ ****************************************************************************/
+
+int chmod(FAR const char *path, mode_t mode)
+{
+  if (path == NULL)
+    {
+      return -EINVAL;
+    }
+
+#ifdef CONFIG_FILE_MODE
+  struct inode_search_s desc;
+  FAR struct inode *inode;
+  int ret;
+
+  /* Get an inode for this file */
+
+  SETUP_SEARCH(&desc, path, false);
+
+  ret = inode_find(&desc);
+  if (ret < 0)
+    {
+      /* "O_CREAT is not set and the named file does not exist.  Or, a
+       * directory component in pathname does not exist or is a dangling
+       * symbolic link."
+       */
+
+      RELEASE_SEARCH(&desc);
+      set_errno(ret);
+      return ERROR;
+    }
+
+  /* Get the search results */
+
+  inode = desc.node;
+  DEBUGASSERT(inode != NULL);
+
+  inode->i_mode = mode;
+
```

Comment:
Shouldn't the mode be restricted for a read-only file system?  Would be an error to set O_WRONLY on a read-only file system?

Reply:
I don't quite understand what you mean. Do you mean that the permission of the file is read-only?

Reply:
> 
> 
> I don't quite understand what you mean. Do you mean that the permission of the file is read-only?

Hmm... I don't think my comment applies to the pseudo file system.  In a true file, the _struct file_operations_ or _struct mountpt_operation_ will have the write method set to NULL on  read only file system.  For example, https://github.com/patacongo/incubator-nuttx/blob/master/fs/romfs/fs_romfs.c#L102

inodes in the pseudo file system on not generally read-able or write-able, but may refer to a read-able or write-able entity such as a device driver or a file (via a mount point).  In those case, I would think that the mode would be handled in a different way.  The file mode is selected for the file or driver via the _open()_ call (or later via _fcntl()).

Shouldn't _chmod()_ work just like _fcntl()_

Reply:
Thank you. You're right. I wrote a little too simple, I will seriously think about what you said.

Reply:
> 
> 
> > I don't quite understand what you mean. Do you mean that the permission of the file is read-only?
> 
> Hmm... I don't think my comment applies to the pseudo file system. In a true file, the _struct file_operations_ or _struct mountpt_operation_ will have the write method set to NULL on read only file system. For example, https://github.com/patacongo/incubator-nuttx/blob/master/fs/romfs/fs_romfs.c#L102
> 
> inodes in the pseudo file system on not generally read-able or write-able, but may refer to a read-able or write-able entity such as a device driver or a file (via a mount point). In those case, I would think that the mode would be handled in a different way. The file mode is selected for the file or driver via the _open()_ call (or later via _fcntl()).
> 
> Shouldn't _chmod()_ work just like _fcntl()_:
> 
> https://github.com/patacongo/incubator-nuttx/blob/master/fs/vfs/fs_fcntl.c#L135
> https://github.com/patacongo/incubator-nuttx/blob/master/fs/vfs/fs_fcntl.c#L163
> 
> Shouldn't _chmod()_  just be a wrappers around _fcntl()_ for real files. I guess I am just having trouble understanding what these APIs mean for the pseudo file system.